mu/modes/esp.py

```diff
@@ -38,15 +38,25 @@ class ESPMode(MicroPythonMode):
     icon = "esp"
     fs = None
 
-    # There are many boards which use ESP microcontrollers but they often use
-    # the same USB / serial chips (which actually define the Vendor ID and
-    # Product ID for the connected devices.
+    # The below list defines the supported devices, however, many
+    # devices are using the exact same FTDI USB-interface, with vendor
+    # ID 0x403 without reporting their own VID/PID
+
+    # In some instances we can recognize the device not on VID/PID,
+    # but on manufacturer ID, that's what the third column is for.
+    # These more specific device specifications, should be listed
+    # before the generic FTDI VID/PID's
     valid_boards = [
         # VID  , PID,    Manufacturer string, Device name
         (0x1A86, 0x7523, None, "HL-340"),
         (0x10C4, 0xEA60, None, "CP210x"),
-        (0x0403, 0x6015, None, "Sparkfun ESP32 Thing"),
         (0x0403, 0x6001, "M5STACK Inc.", "M5Stack ESP32 device"),
+        (0x0403, 0x6001, None, None),  # FT232/FT245 (XinaBox CW01, CW02)
+        (0x0403, 0x6010, None, None),  # FT2232C/D/L/HL/Q (ESP-WROVER-KIT)
```

Comment:
I think the main issue with this case was that the same device creates 2 virtual serial ports, and both will of course contain the same VID/PID pair.

@dybber  I believe with the latest changes in master two devices would appear in the dropdown with the same name, is that right? So a user could switch between the two of them, even if they look identical?

Reply:
I actually just checked, and seems like that is still an issue. Will investigate, the intention is that two identical devices, each of them will show up in the menu, with the same name, but different ports will be shown when you hover over them.

Reply:
I was wrong just before, it does work as intended. Two devices show up:

![Screenshot 2020-09-05 at 19 12 03](https://user-images.githubusercontent.com/55204/92310157-dfb74f00-efab-11ea-9166-513dfeb7b55e.png)

And you can determine which port by hovering:
![Screenshot 2020-09-05 at 19 12 47](https://user-images.githubusercontent.com/55204/92310161-e80f8a00-efab-11ea-991b-9154fe665f97.png)

Reply:
Perhaps it should say "ESP MicroPython device" instead? Currently when the name is `None`, it defaults to the mode name. It could be mode name with `" device"` appended.

Reply:
Yeah, "device" would make it more clear, but maybe it would be more useful to have the port name in brackets? To be able to tell devices apart.
Although we would have to shorten non-Windows ports, `COMXX` is short but the unix paths or port names can be quite long.

Reply:
Yes, I guess when multiple devices of the same type are connected, it would help to differentiate them, without having to mouseover. Teaching wise, it's probably best to just show the entire port-name, instead of trying to hide parts of it. That would just introduce confusion.

Reply:
I will take a look at this when I can find the time, perhaps friday evening

Reply:
Just as a suggestion. We have a flashing tool we use to load precompiled source onto our CW01 and CW02. The tool is also done in Python. We use ESPTOOL, which as you know also is in Python, to flash the code, but we first use the tool to inquire if its a ESP8266 or a ESP32 plugged in, and we show the MAC address at the same time:
<img width="482" alt="Screenshot 2020-09-09 at 12 20 40 PM" src="https://user-images.githubusercontent.com/2776277/92586786-066cd280-f297-11ea-89e6-94bb471407ec.png">

We also use, as shown in this example, the tool to load MicroPython to the xChips.

Reply:
I was thinking a string in the dropdown as `ESP MicroPython (/dev/cu.usbmodem1234567)` might be too long, but maybe that's fine if it helps avoid confusion.

I'll create a ticket to track the work and mark this as resolved, thanks Martin!

Reply:
Ok. How do you detect that it's a CW02 chip, when it just reuses FTDI's VID/PID (0x0403, 0x6001), and it doesn't have its own?

Reply:
It will perhaps be even longer, if it becomes: `ESP MicroPython device (Port: /dev/tty/cu.usbmodem1234567)`, but let's test the idea.

Reply:
I guess on connection we can query the REPL to get the device info, but that would stop the user code every time the device is "detected".

Reply:
> Ok. How do you detect that it's a CW02 chip, when it just reuses FTDI's VID/PID (0x0403, 0x6001), and it doesn't have its own?

Again, just a suggestion, if you want to know which ESP connected to the USB/UART controller: We simply fire off a `python -m esptool --port /dev/cu.usbserial-<actual-port> read_flash` and then scan for `Detecting chip type... ESP32`. We are _not_ determining whether it is a CW01 or CW02 based on VID/PID, we only use that t check that a FTDI chip is connected to the USB port, so the above suggestion is after that. But as I said, just a suggestion. I am happy that it just say ESP MicroPython, but you had an issue trying to determine what was inserted more specifically and this could help.

Reply:
Yeah, taking in consideration the long delays in reading the flash or querying the REPL it sounds like its fine to identify the devices as generic ESP MicroPython.
Thanks again everybody for the clarifications and discussion!